<img src="./airbnb.png" height="96" width="96">
<br />
<div style="text-align: center;">
    <span style="font-weight: bold; color:#FF7674; font-family: 'Courier New'; font-size: 5em;">Madrid Data Cleaning</span>
</div>

<br />
<div style="text-align: left;">
    <span style="color:#8585BB; font-family: 'Courier New'; font-size: 1.5em;">En las siguientes líneas de código estaremos haciendo una limpieza y exploración de los datos obtenidos de www.insideairbnb.com para la ciudad de Madrid.</span>
</div>

In [1]:
import sys
import pandas as pd                   
import datetime as dt                  
import numpy as np                    
import re

pd.set_option('display.max_rows', None)
pd.set_option('display.max_columns', None)

print('Python version: ', sys.version)
print('Pandas version: ', pd.__version__)
print('Today: ', dt.date.today())

('Python version: ', '2.7.12 |Anaconda custom (64-bit)| (default, Jun 29 2016, 11:07:13) [MSC v.1500 64 bit (AMD64)]')
('Pandas version: ', u'0.18.1')
('Today: ', datetime.date(2017, 7, 14))


<div style="text-align: left;">
    <span style="color:#2467C0; font-family: 'Courier New'; font-size: 1.5em;">Vamos a echar un vistazo a nuestros datasets obtenidos de www.insideairbnb.com<br>
La data obtenida es del 08 de abril de 2017</span>
</div>

<div style="text-align: left;">
    <span style="color:#2467C0; font-family: 'Courier New'; font-size: 1.5em;">Según Insideairbnb.com, la siguiente tabla es: *Detailed Listings data for Madrid*</span>
</div>



In [2]:
listing = pd.read_csv('./listings_1.csv', sep=',')
print 'Dimensión:', listing.shape
listing.tail(3)

Dimensión: (13335, 95)


id                            listing_url       scrape_id  \
13332  16392617  https://www.airbnb.com/rooms/16392617  20170407214119   
13333  17840364  https://www.airbnb.com/rooms/17840364  20170407214119   
13334  17829874  https://www.airbnb.com/rooms/17829874  20170407214119   

      last_scraped                                             name  \
13332   2017-04-08  Piso con Terraza 2 personas, Aceptamos Mascotas   
13333   2017-04-08                              Apartamento pozuelo   
13334   2017-04-08                      HABITACIÓN CON BAÑO PRIVADO   

                                                 summary  \
13332  Piso acogedor, habitación individual para los ...   
13333  Apartamento   tranquilo muy cerca de Madrid ca...   
13334  Habitación con baño privado. Zona verde y tran...   

                                                   space  \
13332  Hola! Somos dos hermanas jóvenes y compartimos...   
13333                                                NaN   
13334                                                NaN   

                                             description experiences_offered  \
13332  Piso acogedor, habitación individual para los ...                none   
13333  Apartamento   tranquilo muy cerca de Madrid ca...                none   
13334  Habitación con baño privado. Zona verde y tran...                none   

                                   neighborhood_overview notes transit  \
13332  Está justo al lado del metro Jarama y autobuse...   NaN     NaN   
13333                                                NaN   NaN     NaN   
13334                                                NaN   NaN     NaN   

                                                  access  \
13332  Baño compartido recién reformado en color pist...   
13333                                                NaN   
13334  Habitación con baño . Zonas de acceso a la mis...   

                                             interaction  \
13332  Pueden encontrarme en casa fuera de mi horario...   
13333                                                NaN   
13334                                 Poca conversación.   

                                             house_rules  \
13332  - Hora de entrada y salida Flexible previo avi...   
13333   El precio  incluye  limpieza   De el apartamento   
13334  - No dar mucha conversación ni invadir espacio...   

                                           thumbnail_url  \
13332  https://a0.muscache.com/im/pictures/c9c35abc-f...   
13333  https://a0.muscache.com/im/pictures/1ec37eca-a...   
13334  https://a0.muscache.com/im/pictures/884f1919-7...   

                                              medium_url  \
13332  https://a0.muscache.com/im/pictures/c9c35abc-f...   
13333  https://a0.muscache.com/im/pictures/1ec37eca-a...   
13334  https://a0.muscache.com/im/pictures/884f1919-7...   

                                             picture_url  \
13332  https://a0.muscache.com/im/pictures/c9c35abc-f...   
13333  https://a0.muscache.com/im/pictures/1ec37eca-a...   
13334  https://a0.muscache.com/im/pictures/884f1919-7...   

                                          xl_picture_url    host_id  \
13332  https://a0.muscache.com/im/pictures/c9c35abc-f...   17961354   
13333  https://a0.muscache.com/im/pictures/1ec37eca-a...  107332075   
13334  https://a0.muscache.com/im/pictures/884f1919-7...  121929599   

                                          host_url host_name  host_since  \
13332   https://www.airbnb.com/users/show/17961354     Laila  2014-07-11   
13333  https://www.airbnb.com/users/show/107332075     Chely  2016-12-15   
13334  https://www.airbnb.com/users/show/121929599       Agm  2017-03-21   

                            host_location host_about host_response_time  \
13332  Madrid, Community of Madrid, Spain        NaN       within a day   
13333                                  ES        NaN     within an hour   
13334                                  ES        NaN     within an hou

In [3]:
listing.columns

Index([u'id', u'listing_url', u'scrape_id', u'last_scraped', u'name',
       u'summary', u'space', u'description', u'experiences_offered',
       u'neighborhood_overview', u'notes', u'transit', u'access',
       u'interaction', u'house_rules', u'thumbnail_url', u'medium_url',
       u'picture_url', u'xl_picture_url', u'host_id', u'host_url',
       u'host_name', u'host_since', u'host_location', u'host_about',
       u'host_response_time', u'host_response_rate', u'host_acceptance_rate',
       u'host_is_superhost', u'host_thumbnail_url', u'host_picture_url',
       u'host_neighbourhood', u'host_listings_count',
       u'host_total_listings_count', u'host_verifications',
       u'host_has_profile_pic', u'host_identity_verified', u'street',
       u'neighbourhood', u'neighbourhood_cleansed',
       u'neighbourhood_group_cleansed', u'city', u'state', u'zipcode',
       u'market', u'smart_location', u'country_code', u'country', u'latitude',
       u'longitude', u'is_location_exact', u'prope

<div style="text-align: left;">
    <span style="color:#2467C0; font-family: 'Courier New'; font-size: 1.5em;">Según www.insideairbnb.com, la siguiente tabla es: *Summary information and metrics for listings in Madrid (good for visualisations)*.</span>
</div>

In [4]:
listing_sum = pd.read_csv('./listings.csv', sep=',')
print 'Dimensión:', listing_sum.shape
listing_sum.head(3)

Dimensión: (13335, 16)


id                                         name   host_id  \
0   7830063                    Quiet room in Plaza Mayor  41032693   
1   9898596      Homely apartment in the heart of Madrid  50908516   
2  15334645  Piso Muy Luminoso en pleno centro de Madrid  97753616   

     host_name neighbourhood_group neighbourhood   latitude  longitude  \
0  Bluekitchen              Centro       Palacio  40.412275  -3.708718   
1       Daniel              Centro       Palacio  40.411093  -3.708985   
2         Will              Centro       Palacio  40.413587  -3.708945   

         room_type  price  minimum_nights  number_of_reviews last_review  \
0     Private room     42               2                100  2017-04-02   
1  Entire home/apt    135               1                 68  2017-04-04   
2  Entire home/apt     81               1                 34  2017-03-21   

   reviews_per_month  calculated_host_listings_count  availability_365  
0               5.04                               7               165  
1               4.39                               1               230  
2               5.57                               1               287

<div style="text-align: left;">
    <span style="color:#2467C0; font-family: 'Courier New'; font-size: 1.5em;">Según Insideairbnb.com, la siguiente tabla es: *Detailed Calendar Data for listings in Madrid*.</span>
</div>

In [5]:
calendar = pd.read_csv('./calendar.csv', sep=',')
calendar.tail(3)

listing_id        date available   price
4867272    17829874  2017-04-10         t  $15.00
4867273    17829874  2017-04-09         t  $15.00
4867274    17829874  2017-04-08         f     NaN

In [6]:
#Según Insideairbnb.com, la siguiente tabla es: Detailed Review Data for listings in Madrid
review_sum = pd.read_csv('./reviews_1.csv')
review_sum.head(3)

listing_id        date
0    15128601  2016-09-25
1    15128601  2016-10-03
2    15128601  2016-10-07

In [7]:
#Según Insideairbnb.com, la siguiente tabla es: Summary Review data and Listing ID (to facilitate time based analytics and visualisations linked to a listing).
review = pd.read_csv('./reviews.csv')
review.head(3)

listing_id         id        date  reviewer_id reviewer_name  \
0    15128601  103956852  2016-09-25     48832799        Amparo   
1    15128601  105883025  2016-10-03     87158075       Daniela   
2    15128601  106595943  2016-10-07     10102108        Fabien   

                                            comments  
0  Apartamento muy acogedor, limpio y cómodo, en ...  
1  La descripción del anuncio era correcta, las f...  
2  Nice and cosy place. Really peaceful and quiet...

<div style="text-align: left;">
    <span style="color:#2467C0; font-family: 'Courier New'; font-size: 1.5em;">Una vez vistos los datos que podemos obtener de Inside Airbnb, vamos a descartar las tablas resumidas y usar las tablas que tienen mayor cantidad de datos para poder moldearlas a nuestro gusto.<br><br></span>
</div>

<div style="text-align: left;">
    <span style="color:#2467C0; font-family: 'Courier New'; font-size: 1.5em;">Las tablas que finalmente utilizaremos serán:<br><br>
- listing = pd.read_csv('./listings_1.csv', sep=',')<br>
- calendar = pd.read_csv('./calendar.csv', sep=',')<br>
- review = pd.read_csv('./reviews.csv')</span>
</div>

<br><div style="text-align: left;">
    <span style="font-weight: bold; color:#FF7674; font-family: 'Courier New'; font-size: 2em;">Data Cleaning</span>
</div>
<div style="text-align: left;">
    <span style="color:#2467C0; font-family: 'Courier New'; font-size: 1.5em;">Aunque los datos reales nunca están limpios completamente, necesitamos asegurarnos de limpiarlos lo mayor posible convirtiendo o eliminando valores nulos o faltantes.<br><br></span>
</div>


In [8]:
listing.describe().transpose()

C:\Users\ronny\Anaconda2\lib\site-packages\numpy\lib\function_base.py:3834: RuntimeWarning: Invalid value encountered in percentile
  RuntimeWarning)


count          mean           std  \
id                              13335.0  1.043929e+07  5.517349e+06   
scrape_id                       13335.0  2.017041e+13  4.890808e+00   
host_id                         13335.0  3.775842e+07  3.470057e+07   
host_acceptance_rate                0.0           NaN           NaN   
host_listings_count             13332.0  9.777453e+00  2.781298e+01   
host_total_listings_count       13332.0  9.777453e+00  2.781298e+01   
latitude                        13335.0  4.042073e+01  2.034312e-02   
longitude                       13335.0 -3.697255e+00  2.383740e-02   
accommodates                    13335.0  3.184252e+00  1.990534e+00   
bathrooms                       13285.0  1.251675e+00  5.981173e-01   
bedrooms                        13312.0  1.293495e+00  8.247503e-01   
beds                            13286.0  1.981936e+00  1.511608e+00   
square_feet                       522.0  3.758352e+02  5.459670e+02   
guests_included                 13335.0  1.572628e+00  1.088805e+00   
minimum_nights                  13335.0  3.009824e+00  1.282894e+01   
maximum_nights                  13335.0  9.597532e+02  8.883119e+03   
has_availability                    0.0           NaN           NaN   
availability_30                 13335.0  8.831496e+00  9.152567e+00   
availability_60                 13335.0  2.307762e+01  1.958616e+01   
availability_90                 13335.0  4.030664e+01  2.942652e+01   
availability_365                13335.0  2.058828e+02  1.269678e+02   
number_of_reviews               13335.0  2.298755e+01  3.839320e+01   
review_scores_rating            10417.0  9.162062e+01  8.983817e+00   
review_scores_accuracy          10399.0  9.408020e+00  9.244271e-01   
review_scores_cleanliness       10405.0  9.326382e+00  9.949704e-01   
review_scores_checkin           10389.0  9.622293e+00  7.916267e-01   
review_scores_communication     10405.0  9.644786e+00  7.575793e-01   
review_scores_location          10387.0  9.542794e+00  7.666081e-01   
review_scores_value             10387.0  9.212573e+00  9.541097e-01   
jurisdiction_names                  0.0           NaN           NaN   
calculated_host_listings_count  13335.0  7.600075e+00  1.961052e+01   
reviews_per_month               10542.0  1.923220e+00  1.888180e+00   

                                         min           25%           50%  \
id                              1.862800e+04  5.858327e+06  1.157754e+07   
scrape_id                       2.017041e+13  2.017041e+13  2.017041e+13   
host_id                         1.745300e+04  7.823688e+06  2.752699e+07   
host_acceptance_rate                     NaN           NaN           NaN   
host_listings_count             0.000000e+00           NaN           NaN   
host_total_listings_count       0.000000e+00           NaN           NaN   
latitude                        4.033189e+01  4.041009e+01  4.041872e+01   
longitude                      -3.863907e+00 -3.707852e+00 -3.701596e+00   
accommodates                    1.000000e+00  2.000000e+00  2.000000e+00   
bathrooms                       0.000000e+00           NaN           NaN   
bedrooms                        0.000000e+00           NaN           NaN   
beds                            1.000000e+00           NaN           NaN   
square_feet                     0.000000e+00           NaN           NaN   
guests_included                 1.000000e+00  1.000000e+00  1.000000e+00   
minimum_nights                  1.000000e+00  1.000000e+00  2.000000e+00   
maximum_nights                  1.000000e+00  3.650000e+02  1.125000e+03   
has_availability                         NaN           NaN           NaN   
availability_30                 0.000000e+00  0.000000e+00  6.000000e+00   
availability_60                 0.000000e+00  5.000000e+00  2.000000e+01   
availability_90                 0.000000e+00  1.300000e+01  3.800000e+01   
availability_365                0.000000e+00  8.400000e+01  2.470000e+02   
number_of_reviews 

In [9]:
listing.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13335 entries, 0 to 13334
Data columns (total 95 columns):
id                                  13335 non-null int64
listing_url                         13335 non-null object
scrape_id                           13335 non-null int64
last_scraped                        13335 non-null object
name                                13335 non-null object
summary                             12846 non-null object
space                               9700 non-null object
description                         13329 non-null object
experiences_offered                 13335 non-null object
neighborhood_overview               8308 non-null object
notes                               5037 non-null object
transit                             8264 non-null object
access                              7546 non-null object
interaction                         7523 non-null object
house_rules                         8666 non-null object
thumbnail_url                 

In [10]:
#Existe alguna fila NULL?
print listing.isnull().any().any() 
print listing.shape

True
(13335, 95)


<div style="text-align: left;">
    <span style="color:#2467C0; font-family: 'Courier New'; font-size: 1.5em;">Vamos a tratar de encontrar cuántos valores nulos existen en cada columna!</span>
</div>


In [11]:
listing.isnull().sum(axis=0)

id                                      0
listing_url                             0
scrape_id                               0
last_scraped                            0
name                                    0
summary                               489
space                                3635
description                             6
experiences_offered                     0
neighborhood_overview                5027
notes                                8298
transit                              5071
access                               5789
interaction                          5812
house_rules                          4669
thumbnail_url                        2488
medium_url                           2488
picture_url                             0
xl_picture_url                       2488
host_id                                 0
host_url                                0
host_name                               3
host_since                              3
host_location                     

<div style="text-align: left;">
    <span style="color:#2467C0; font-family: 'Courier New'; font-size: 1.5em;">El paso anterior ya nos dió un indicio de que hay varias columnas con datos únicamente nulos.</span>
</div>

<br><div style="text-align: left;">
    <span style="font-weight: bold; color:#FF7674; font-family: 'Courier New'; font-size: 2em;">Arreglando los Missing Values</span>
</div>
<div style="text-align: left;">
    <span style="color:#2467C0; font-family: 'Courier New'; font-size: 1.5em;">Primero vamos a borrar las columnas cuyos valores sean todos Nulos, ya que no aportan nada. Una de las posibles causas puede haber sido una mala captura de datos.<br><br></span>
</div>

In [12]:
listing.isnull().all()

id                                  False
listing_url                         False
scrape_id                           False
last_scraped                        False
name                                False
summary                             False
space                               False
description                         False
experiences_offered                 False
neighborhood_overview               False
notes                               False
transit                             False
access                              False
interaction                         False
house_rules                         False
thumbnail_url                       False
medium_url                          False
picture_url                         False
xl_picture_url                      False
host_id                             False
host_url                            False
host_name                           False
host_since                          False
host_location                     

In [13]:
print listing['jurisdiction_names'].unique()
print listing['has_availability'].unique()
print listing['host_acceptance_rate'].unique()

[ nan]
[ nan]
[ nan]


In [14]:
listing.drop('jurisdiction_names', axis=1, inplace=True)
listing.drop('has_availability', axis=1, inplace=True)
listing.drop('host_acceptance_rate', axis=1, inplace=True)

In [15]:
listing.shape

(13335, 92)

<br><div style="text-align: left;">
    <span style="font-weight: bold; color:#FF7674; font-family: 'Courier New'; font-size: 2em;">Sampling</span>
</div>

<div style="text-align: left;">
    <span style="color:#2467C0; font-family: 'Courier New'; font-size: 1.5em;">Nuestra tabla de datos tiene muchas líneas o registros. Para no afectar el funcionamiento de nuestro ordenador, cuando queramos revisar los tipos de datos sólo miraremos pocas líneas a la vez. Ahora bien, en vez de mirar las mismas 2, 5 ó 10 líneas principales cada vez, aplicaremos lo siguiente con lo que conseguiremos ver una muestra aleatoria diferente en nuestro head(). De esta manera, nos aseguramos de que los datos no estén en ningún orden particular cuando probamos el muestreo (como tomar filas de arriba o de abajo) al mezclar aleatoriamente las filas.<br><br></span>
</div>

In [16]:
#Shuffle de las filas de "listing" para que obtengamos una muestra distribuida cuando mostramos el head()

listing = listing.reindex(np.random.permutation(listing.index))
listing.head(3)

id                            listing_url       scrape_id  \
8965   18095441  https://www.airbnb.com/rooms/18095441  20170407214119   
5575    7745816   https://www.airbnb.com/rooms/7745816  20170407214119   
11498   7469234   https://www.airbnb.com/rooms/7469234  20170407214119   

      last_scraped                                               name  \
8965    2017-04-08  Habitación con cama de matrimonio Barrio Salam...   
5575    2017-04-08                 Precioso apto. al lado de Gran Vía   
11498   2017-04-08           MADRID RIO room, luminous, metro station   

                                                 summary  \
8965   Habitación privada con cama de matrimonio. Der...   
5575   Apartamento muy acogedor completamente reforma...   
11498  MADRID RIO. Beautiful and bright room. 20 mete...   

                                                   space  \
8965                  La ubicación, limpieza y comodidad   
5575   El apartamento consta de dos espacios, como se...   
11498           Magnifient apartment, great common areas   

                                             description experiences_offered  \
8965   Habitación privada con cama de matrimonio. Der...                none   
5575   Apartamento muy acogedor completamente reforma...                none   
11498  MADRID RIO. Beautiful and bright room. 20 mete...                none   

                                   neighborhood_overview  \
8965                                               Bueno   
5575   Localización inmejorable. El barrio es el míti...   
11498  5 minutes walking from Madrid Rio y and Princi...   

                                                   notes  \
8965   Recomendado para viajes de negocios, placer ó ...   
5575                                                 NaN   
11498                                Car parking 12€/day   

                                                 transit  \
8965   Entrada del Metro de O'Donell, Goya y Manuel B...   
5575   Hay varios metros al lado de casa: Noviciado, ...   
11498  The closest metro station is less than 20 mete...   

                                                  access  \
8965   Cocina-comedor y baño a compartir con propietaria   
5575   Al ser una vivienda de uso propio que se ofrec...   
11498       Living room/dining room. Internet WiFi 20Mb.   

                                             interaction  \
8965   Sólo se coincidirá puntuslmente a lo largo del...   
5575   Dispondréis de mi teléfono por si tuvierais cu...   
11498  The keys are delivered in person. Facilities a...   

                                             house_rules  \
8965   Respetar las horas de silencio en la noche, y ...   
5575   El apartamento se entrega limpio y en perfecto...   
11498  Enjoy the apartment and respec it. Respect to ...   

                                           thumbnail_url  \
8965   https://a0.muscache.com/im/pictures/e848ba24-d...   
5575                                                 NaN   
11498                                                NaN   

                                              medium_url  \
8965   https://a0.muscache.com/im/pictures/e848ba24-d...   
5575                                                 NaN   
11498                                                NaN   

                                             picture_url  \
8965   https://a0.muscache.com/im/pictures/e848ba24-d...   
5575   https://a0.muscache.com/im/pictures/98576927/7...   
11498  https://a0.muscache.com/im/pictures/94700232/4...   

                                          xl_picture_url    host_id  \
8965   https://a0.muscache.com/im/pictures/e848ba24-d...  124597861   
5575                                                 NaN   19227342   
11498                                                NaN   24085430   

                                          host_url host_name  host_since  \
8965   https://www.airbnb.com/users/show/124597861    Esther  2017-04-06   
5575    https:

<div style="text-align: left;">
    <span style="color:#2467C0; font-family: 'Courier New'; font-size: 1.5em;">Existen otras columnas con datos que no nos interesan, por ellos procederemos a eliminarlas:
<br></span>
</div>

In [17]:
delete = ['listing_url', 'scrape_id', 'last_scraped', 'thumbnail_url', 'medium_url','country_code','calendar_last_scraped', 
          'is_location_exact', 'host_neighbourhood', 'picture_url', 'xl_picture_url', 'host_thumbnail_url', 'host_picture_url', 
          'summary', 'description','host_url', 'experiences_offered', 'space', 'neighborhood_overview', 'notes', 'transit', 
          'access', 'interaction', 'house_rules', 'host_about', 'host_response_time', 'host_response_rate', 'square_feet', 
          'license']   
len(delete)

29

<div style="text-align: left;">
    <span style="color:#2467C0; font-family: 'Courier New'; font-size: 1.5em;">El dato de "square_feet" lo borramos porque contiene muchos datos nulos, aunque podría ser un dato interesante para -por ejemplo- determinar el precio de alquiler del piso; aunque en el caso de AirBnB (si vamos a la concepción inicial de la idea de la compañia: compartir sofá) también podría considerarse como un dato menor. 
<br></span>
</div>

In [18]:
for i in range(0,len(delete)):
    listing.drop(delete[i], axis=1, inplace=True)

In [19]:
listing.shape

(13335, 63)

In [20]:
listing = listing.reindex(np.random.permutation(listing.index))
listing.head(3)

id                                 name   host_id       host_name  \
9135   1690312  Apartamento 200m2 para 5/6 personas   8815003           Maite   
441    4678105   Central sunny apartment with views  17705919  Barbara Y Javi   
5021  13091807  long rental study old town madrid 2  21137168             Ale   

      host_since                       host_location host_is_superhost  \
9135  2013-09-13  Madrid, Community of Madrid, Spain                 f   
441   2014-07-06  Madrid, Community of Madrid, Spain                 f   
5021  2014-09-09                              Spain                  f   

      host_listings_count  host_total_listings_count  \
9135                  2.0                        2.0   
441                   1.0                        1.0   
5021                  4.0                        4.0   

                          host_verifications host_has_profile_pic  \
9135  ['email', 'phone', 'reviews', 'jumio']                    t   
441   ['email', 'phone', 'reviews', 'jumio']                    t   
5021           ['email', 'phone', 'reviews']                    t   

     host_identity_verified  \
9135                      t   
441                       t   
5021                      f   

                                                 street neighbourhood  \
9135           Madrid, Community of Madrid 28016, Spain           NaN   
441            Madrid, Comunidad de Madrid 28005, Spain           NaN   
5021  Malasaña, Madrid, Comunidad de Madrid 28004, S...      Malasaña   

     neighbourhood_cleansed neighbourhood_group_cleansed    city  \
9135         Hispanoamérica                    Chamartín  Madrid   
441                 Palacio                       Centro  Madrid   
5021            Universidad                       Centro  Madrid   

                    state zipcode  market smart_location country   latitude  \
9135  Community of Madrid   28016  Madrid  Madrid, Spain   Spain  40.456181   
441   Comunidad de Madrid   28005  Madrid  Madrid, Spain   Spain  40.412441   
5021  Comunidad de Madrid   28004  Madrid  Madrid, Spain   Spain  40.426694   

      longitude property_type        room_type  accommodates  bathrooms  \
9135  -3.676956     Apartment  Entire home/apt             6        2.0   
441   -3.711110     Apartment  Entire home/apt             2        1.0   
5021  -3.707727     Apartment     Private room             2        1.0   

      bedrooms  beds  bed_type  \
9135       3.0   3.0  Real Bed   
441        2.0   2.0  Real Bed   
5021       2.0   1.0  Real Bed   

                                              amenities    price weekly_price  \
9135  {TV,"Cable TV",Internet,"Wireless Internet",Ki...  $200.00          NaN   
441   {TV,Internet,"Wireless Internet",Kitchen,"Elev...  $100.00          NaN   
5021  {Internet,"Wireless Internet","Air conditionin...   $39.00          NaN   

     monthly_price security_deposit cleaning_fee  guests_included  \
9135           NaN          $500.00          NaN                1   
441            NaN          $150.00       $10.00                1   
5021           NaN              NaN          NaN                1   

     extra_people  minimum_nights  maximum_nights calendar_updated  \
9135       $50.00               2               7      2 weeks ago   
441         $0.00               2            1125      2 weeks ago   
5021        $0.00               7             365            today   

      availability_30  availability_60  availability_90  availability_365  \
9135               30               60               90               365   
441                 0                6               36               311   
5021                7                9               29               281   

      number_of_reviews first_review last_review  review_scores_rating  \
9135                  1   2015-05-24  2015-05-24                  80.0   
441                  20   2015-02-01  2016-10-04                  94.0   
5021                  2   20

In [21]:
listing.price.head()

9135     $200.00
441      $100.00
5021      $39.00
12574    $120.00
2928      $60.00
Name: price, dtype: object

<div style="text-align: left;">
    <span style="color:#2467C0; font-family: 'Courier New'; font-size: 1.5em;">Debo eliminar el simbolo de dolar en la columna de precio y convertirlo en un dato tipo float.<br></span>
</div>

In [22]:
def transform_price(df, variables=None):
    df = df.copy()
    for variable in variables:
        df[variable] = df[variable].str.strip('$').str.replace(',', '') \
            # Remove dollar signs and thousands separators
        df[variable] = df[variable].astype(np.float64)
    return df


vars_price = ['cleaning_fee', 'price', 'extra_people', 'weekly_price', 'monthly_price' ]
listing = transform_price(listing, vars_price)

In [23]:
listing['price'].head(5)

9135     200.0
441      100.0
5021      39.0
12574    120.0
2928      60.0
Name: price, dtype: float64

<div style="text-align: left;">
    <span style="color:#2467C0; font-family: 'Courier New'; font-size: 1.5em;">También se podría hacer de la siguiente manera:<br><br></span></div>
    
<div style="text-align: left;">
    <span style="color:#2467C0; font-family: 'Courier New'; font-size: 1.2em;"> listing['price']=listing['price'].str.strip('$').str.replace(',', '') <br>
listing['price'] = pd.to_numeric(listing.price)</span></div>

<div style="text-align: left;">
    <span style="color:#2467C0; font-family: 'Courier New'; font-size: 1.5em;">O tambien, quitamos el símbolo de $ y el separador de miles en los valores no nulos "NA":
<br><br></span></div>

<div style="text-align: left;">
    <span style="color:#2467C0; font-family: 'Courier New'; font-size: 1.2em;">listing['price'] = listing.loc[listing.price.notnull()].price.apply(lambda x: re.sub('$', '', x))<br>
listing['price'] = pd.to_numeric(listing.price)
<br></span></div>



<div style="text-align: left;">
    <span style="color:#2467C0; font-family: 'Courier New'; font-size: 1.5em;">Seguimos limpiando datos...<br></span>
</div>

In [24]:
variables = ['state', 'city', 'market', 'smart_location', 'country', 'neighbourhood', 'neighbourhood_cleansed',
             'neighbourhood_group_cleansed', 'host_location', 'street']

for i in range(0,len(variables)-1):
    print variables[i]
    print listing[variables[i]].unique()
    print ''

state
['Community of Madrid' 'Comunidad de Madrid' nan 'Madrid'
 '\xe9\xa6\xac\xe5\xbe\xb7\xe9\x87\x8c\xe8\x87\xaa\xe6\xb2\xbb\xe5\x8d\x80'
 'madrid'
 '\xe9\xa9\xac\xe5\xbe\xb7\xe9\x87\x8c\xe8\x87\xaa\xe6\xb2\xbb\xe5\x8c\xba'
 'Communaut\xc3\xa9 de Madrid' 'Madrid \nMadrid' 'Salerno Fisciano' 'Spain'
 'Madrid, Spain' 'Espa\xc3\xb1a' 'Autonome Gemeinschaft Madrid'
 '\xe9\xa9\xac\xe5\xbe\xb7\xe9\x87\x8c' '28' 'Usera' 'Valencian Community'
 'MADRID' 'Madrid Espa\xc3\xb1a' ' ' 'Alicante' '  ' 'Espa\xc5\x84a'
 'Madrid capital' 'La Habana' 'Barcelona'
 '\xe8\xa5\xbf\xe7\x8f\xad\xe7\x89\x99' 'Espa\xc3\xb1a,Madrid'
 'Regi\xc3\xa3o Aut\xc3\xb4noma de Madrid']

city
['Madrid' '\xe9\xa6\xac\xe5\xbe\xb7\xe9\x87\x8c'
 'Madrid, Comunidad de Madrid, ES' '\xe9\xa9\xac\xe5\xbe\xb7\xe9\x87\x8c'
 'Madrid ' 'Valderrodrigo' ' Madrid' 'Salerno' 'Pozuelo de Alarc\xc3\xb3n'
 'San L\xc3\xa1zaro' 'M\xc3\xa1laga' 'Boadilla del Monte' 'Pozuelo'
 'madrid' 'MADRID' 'MADRID ' 'Chueca, Madrid' nan 'villa verde bajo'


In [25]:
state = ['Community of Madrid', '\xe9\xa9\xac\xe5\xbe\xb7\xe9\x87\x8c\xe8\x87\xaa\xe6\xb2\xbb\xe5\x8c\xba',
         'Regi\xc3\xa3o Aut\xc3\xb4noma de Madrid', 'madrid', '\xe9\xa9\xac\xe5\xbe\xb7\xe9\x87\x8c',
         '\xe9\xa6\xac\xe5\xbe\xb7\xe9\x87\x8c\xe8\x87\xaa\xe6\xb2\xbb\xe5\x8d\x80',
         'Espa\xc3\xb1a', '\xe8\xa5\xbf\xe7\x8f\xad\xe7\x89\x99',
         'Espa\xc3\xb1a,Madrid', 'MADRID', 'Espa\xc5\x84a',
         'Madrid capital', 'Spain', 'Communaut\xc3\xa9 de Madrid',
         'Madrid \nMadrid', 'Madrid Espa\xc3\xb1a', 'Madrid, Spain',
         'Autonome Gemeinschaft Madrid', 'Usera', ' ', '  ', '28','Comunidad de Madrid','']

for i in range(0, len(state)-1):
    listing.loc[(listing.state == state[i]), ['state']] = 'Madrid'


In [26]:
listing.state.unique()

array(['Madrid', nan, 'Salerno Fisciano', 'Valencian Community',
       'Alicante', 'La Habana', 'Barcelona'], dtype=object)

In [27]:
listing.state.value_counts()

Madrid                 13287
Salerno Fisciano           1
Barcelona                  1
La Habana                  1
Alicante                   1
Valencian Community        1
Name: state, dtype: int64

<div style="text-align: left;">
    <span style="color:#2467C0; font-family: 'Courier New'; font-size: 1.5em;">En las celdas anteriores verificamos que hay que hacer una labor de limpieza con los datos de las variables que se relacionan con la ciudad de Madrid y deshacernos de los registros que no pertenecen a la capital.<br><br></span></div>

In [28]:
listing = listing[listing.state == 'Madrid']
listing = listing[listing.country == 'Spain']
listing.shape

(13287, 63)

In [29]:
#Limpieza de neighbourhood_group_cleansed
listing.loc[(listing['neighbourhood_group_cleansed'] == 'Tetu\xc3\xa1n'), ['neighbourhood_group_cleansed']] = 'Tetuan'
listing.loc[(listing['neighbourhood_group_cleansed'] == 'Chamart\xc3\xadn'), ['neighbourhood_group_cleansed']] = 'Chamartin'
listing.loc[(listing['neighbourhood_group_cleansed'] == 'Chamber\xc3\xad'), ['neighbourhood_group_cleansed']] = 'Chamberi'
listing.loc[(listing['neighbourhood_group_cleansed'] == 'Vic\xc3\xa1lvaro'), ['neighbourhood_group_cleansed']] = 'Vicalvario'

In [30]:
listing.neighbourhood_group_cleansed.value_counts()

Centro                   6754
Chamberi                  958
Salamanca                 882
Arganzuela                788
Tetuan                    463
Moncloa - Aravaca         445
Retiro                    438
Latina                    379
Chamartin                 360
Carabanchel               357
Ciudad Lineal             309
Puente de Vallecas        222
Hortaleza                 185
Fuencarral - El Pardo     169
Usera                     144
San Blas - Canillejas     117
Villaverde                 83
Barajas                    80
Moratalaz                  74
Villa de Vallecas          46
Vicalvario                 34
Name: neighbourhood_group_cleansed, dtype: int64

<div style="text-align: left;">
    <span style="color:#2467C0; font-family: 'Courier New'; font-size: 1.5em;">Nos vamos dando cuenta de los barrios con más opciones de alquiler en AirBnB.<br><br></span></div>

In [31]:
col_to_delete = ['host_location', 'neighbourhood', 'country','street', 'smart_location','market', 'city', 'state', 'neighbourhood_cleansed']

In [32]:
for i in range(0,len(col_to_delete)):
    listing.drop(col_to_delete[i], axis=1, inplace=True)

listing.shape

(13287, 54)

In [33]:
listing.isnull().sum(axis=0)

id                                     0
name                                   0
host_id                                0
host_name                              3
host_since                             3
host_is_superhost                      3
host_listings_count                    3
host_total_listings_count              3
host_verifications                     0
host_has_profile_pic                   3
host_identity_verified                 3
neighbourhood_group_cleansed           0
zipcode                              439
latitude                               0
longitude                              0
property_type                          0
room_type                              0
accommodates                           0
bathrooms                             49
bedrooms                              23
beds                                  49
bed_type                               0
amenities                              0
price                                  0
weekly_price    

<br><div style="text-align: left;">
    <span style="font-weight: bold; color:#FF7674; font-family: 'Courier New'; font-size: 2em;">¿Existen outliers en mi dataset? ¿Qué precio y características tienen los pisos más costosos de mi conjunto de datos?</span>
</div>
<div style="text-align: left;">
    <span style="color:#2467C0; font-family: 'Courier New'; font-size: 1.5em;">Sería interesante conocer qué precios tienen los pisos más costosos de mi dataset y qué carácteristicas tienen.<br><br></span>
</div>

In [34]:
outliers = listing[listing['price'] >= 1000]
outliers['price'].value_counts()

1000.0    3
3000.0    2
3189.0    1
7700.0    1
1890.0    1
1200.0    1
Name: price, dtype: int64

In [35]:
outliers.sort_values('price')

id                                               name    host_id  \
10667  16199273                 Private Room in Rios Rosas, Madrid  102392514   
5814   10715621                  Lujoso apartamento en Gran Vìa 48    3256859   
957    10495669                  Ático centrico y tranquilo / Atic   46129526   
2246   14574259  LONGTERM welcome-New,Central,100m² for 1-6 Per...   90311483   
8855    6681279                Fantastico Piso Barrio de Salamanca   34972297   
7993     361488                RETIRO PARK2 CHALET INDUSTRIAL-CHIC    1732442   
8002    1727101                RETIRO PARK1 CHALET INDUSTRIAL-CHIC    1732442   
8027    2744574                          Mansion with royal garden    9631554   
10828   1250842                Habitación Moncloa-AravacaArgüelles    6815584   

         host_name  host_since host_is_superhost  host_listings_count  \
10667       Hemant  2016-11-03                 f                  1.0   
5814    Fer & Fran  2012-08-14                 f                 14.0   
957          Elisa  2015-10-09                 f                  1.0   
2246        Tobias  2016-08-16                 f                  1.0   
8855   Juan Andres  2015-06-04                 f                  1.0   
7993         Anouk  2012-02-10                 f                 12.0   
8002         Anouk  2012-02-10                 f                 12.0   
8027           K&S  2013-10-25                 t                  1.0   
10828       Héctor  2013-06-09                 f                  1.0   

       host_total_listings_count                         host_verifications  \
10667                        1.0                         ['email', 'phone']   
5814                        14.0     ['email', 'phone', 'reviews', 'jumio']   
957                          1.0  ['email', 'phone', 'facebook', 'reviews']   
2246                         1.0              ['email', 'phone', 'reviews']   
8855                         1.0                         ['email', 'phone']   
7993                        12.0     ['email', 'phone', 'reviews', 'jumio']   
8002                        12.0     ['email', 'phone', 'reviews', 'jumio']   
8027                         1.0     ['email', 'phone', 'reviews', 'jumio']   
10828                        1.0                                         []   

      host_has_profile_pic host_identity_verified  \
10667                    t                      f   
5814                     t                      t   
957                      t                      f   
2246                     t                      f   
8855                     t                      f   
7993                     t                      t   
8002                     t                      t   
8027                     t                      t   
10828                    t                      f   

      neighbourhood_group_cleansed zipcode   latitude  longitude  \
10667                     Chamberi   28003  40.441251  -3.697699   
5814                        Centro   28013  40.421617  -3.705310   
957                         Centro   28013  40.421374  -3.710629   
2246                        Centro   28005  40.409006  -3.710069   
8855                     Salamanca   28006  40.434467  -3.682262   
7993                        Retiro   28007  40.409084  -3.673131   
8002                        Retiro   28007  40.410194  -3.671795   
8027                     Salamanca       -  40.423505  -3.689261   
10828            Moncloa - Aravaca   28008  40.430939  -3.721799   

      property_type        room_type  accommodates  bathrooms  bedrooms  beds  \
10667     Apartment     Private room             1        1.0       1.0   1.0   
5814      Apartment  Entire home/apt             4        1.0       1.0   1.0   
957       Apartment  Entire home/apt             4        1.0       1.0   2.0   
2246      Apartment  Entire home/apt             6        2.0       2.0   2.0   
8855    Condominium  Entire home/apt             4        2.0  

<div style="text-align: left;">
    <span style="color:#2467C0; font-family: 'Courier New'; font-size: 1.5em;">Podemos sacar varias conclusiones:<br> <br>
    - 1) Existe un alquiler de 7700 Euros por noche, por una habitación! A menos que sea una habitación del Palacio Real, no lo creo muy real. Eliminamos este regitro (id=1250842)<br>
    - 2) Existe otro registro, que es una habitación de alquiler por 1000 euros la noche. Parece no tener mucho sentido. (id=16199273)<br>
    - 3) Existen pisos que en el campo 'number_of_reviews' tienen cero comentarios. Esto indica que nunca han sido alquilados. Tomando en consideración el precio, pueden ser considerados como outliers. Eliminar.<br></span>
</div>


In [36]:
id_to_delete = [6681279, 361488, 1727101, 2744574, 1250842, 16199273]
for i in range(0,len(id_to_delete)):
    listing = listing[listing['id'] != id_to_delete[i] ]

In [37]:
#outliers.dropna(subset=['reviews_per_month'])

<br><br><div style="text-align: left;">
    <span style="color:#2467C0; font-family: 'Courier New'; font-size: 1.5em;">Tenemos otra serie de datos (que aún no sé si utilizaré pero pueden servirme a futuro) que deben ser cambiados a su formato correcto. Es el caso de variables con fechas y variables booleanas, que se tratan a continuación:</span>
</div>

In [38]:
# Cambiamos el formato a booleano
bools = ['host_is_superhost', 'host_has_profile_pic', 'host_identity_verified', 'requires_license', 'instant_bookable', 
         'require_guest_profile_picture', 'require_guest_phone_verification']
for i in bools:
    listing[i] = listing[i].map({'t': True, 'f': False})


In [39]:
# Cambiamos el formato de las fechas
fechas = ['host_since', 'first_review', 'last_review']
for i in fechas:
    listing[i] = pd.to_datetime(listing[i], infer_datetime_format=True)

In [40]:
listing.head(3)

id                                 name   host_id       host_name  \
9135   1690312  Apartamento 200m2 para 5/6 personas   8815003           Maite   
441    4678105   Central sunny apartment with views  17705919  Barbara Y Javi   
5021  13091807  long rental study old town madrid 2  21137168             Ale   

     host_since host_is_superhost  host_listings_count  \
9135 2013-09-13             False                  2.0   
441  2014-07-06             False                  1.0   
5021 2014-09-09             False                  4.0   

      host_total_listings_count                      host_verifications  \
9135                        2.0  ['email', 'phone', 'reviews', 'jumio']   
441                         1.0  ['email', 'phone', 'reviews', 'jumio']   
5021                        4.0           ['email', 'phone', 'reviews']   

     host_has_profile_pic host_identity_verified neighbourhood_group_cleansed  \
9135                 True                   True                    Chamartin   
441                  True                   True                       Centro   
5021                 True                  False                       Centro   

     zipcode   latitude  longitude property_type        room_type  \
9135   28016  40.456181  -3.676956     Apartment  Entire home/apt   
441    28005  40.412441  -3.711110     Apartment  Entire home/apt   
5021   28004  40.426694  -3.707727     Apartment     Private room   

      accommodates  bathrooms  bedrooms  beds  bed_type  \
9135             6        2.0       3.0   3.0  Real Bed   
441              2        1.0       2.0   2.0  Real Bed   
5021             2        1.0       2.0   1.0  Real Bed   

                                              amenities  price  weekly_price  \
9135  {TV,"Cable TV",Internet,"Wireless Internet",Ki...  200.0           NaN   
441   {TV,Internet,"Wireless Internet",Kitchen,"Elev...  100.0           NaN   
5021  {Internet,"Wireless Internet","Air conditionin...   39.0           NaN   

      monthly_price security_deposit  cleaning_fee  guests_included  \
9135            NaN          $500.00           NaN                1   
441             NaN          $150.00          10.0                1   
5021            NaN              NaN           NaN                1   

      extra_people  minimum_nights  maximum_nights calendar_updated  \
9135          50.0               2               7      2 weeks ago   
441            0.0               2            1125      2 weeks ago   
5021           0.0               7             365            today   

      availability_30  availability_60  availability_90  availability_365  \
9135               30               60               90               365   
441                 0                6               36               311   
5021                7                9               29               281   

      number_of_reviews first_review last_review  review_scores_rating  \
9135                  1   2015-05-24  2015-05-24                  80.0   
441                  20   2015-02-01  2016-10-04                  94.0   
5021                  2   2017-03-07  2017-03-20                  80.0   

      review_scores_accuracy  review_scores_cleanliness  \
9135                     8.0                       10.0   
441                      9.0                        9.0   
5021                    10.0                        9.0   

      review_scores_checkin  review_scores_communication  \
9135                   10.0                          8.0   
441                    10.0                         10.0   
5021                   10.0                          9.0   

      review_scores_location  review_scores_value requires_license  \
9135                    10.0                  8.0             True   
441                     10.0                  9.0             True   
5021                    10.0                 10.0             True   

     instant_bookable cancellation_policy require_guest_pr

<div style="text-align: left;">
    <span style="color:#2467C0; font-family: 'Courier New'; font-size: 1.5em;">Generamos el nuevo fichero con la data limpia</span>
</div>

In [41]:
listing.to_csv('./listing_clean.csv', sep=',')

<div style="text-align: left;"><br><br><br>
    <span style="color:#2467C0; font-family: 'Courier New'; font-size: 1.5em;">De momento hemos limpiado los datos del dataset Listing...<br><br>Seguimos...<br><br></span>
</div>

<br><div style="text-align: left;">
    <span style="font-weight: bold; color:#FF7674; font-family: 'Courier New'; font-size: 2em;">Dataset "Calendar":</span>
</div>

<div style="text-align: left;">
    <span style="color:#2467C0; font-family: 'Courier New'; font-size: 1.5em;">
Vamos ahora a trabajar con el dataset Calendar. Este dataset es importante para conocer por ejemplo, la disponibilidad de pisos durante un año, entre otras cosas interesantes.</span>
</div>

In [42]:
calendar.head()

listing_id        date available   price
0     7830063  2018-01-03         t  $42.00
1     7830063  2018-01-02         t  $42.00
2     7830063  2018-01-01         t  $42.00
3     7830063  2017-12-31         f     NaN
4     7830063  2017-12-30         f     NaN

<div style="text-align: left;">
    <span style="color:#2467C0; font-family: 'Courier New'; font-size: 1.5em;">
Revisamos los data types de las columnas para saber si son del tipo correcto</span>
</div>

In [43]:
calendar.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4867275 entries, 0 to 4867274
Data columns (total 4 columns):
listing_id    int64
date          object
available     object
price         object
dtypes: int64(1), object(3)
memory usage: 148.5+ MB


<div style="text-align: left;">
    <span style="color:#2467C0; font-family: 'Courier New'; font-size: 1.5em;">
EL dataset calendar contiene data de los precios para cada piso en alquiler para todo un año y con información de la disponibilidad de los mismos.
<br><br>
Al igual que el dataset anterior, revisamos que los tipos de datos sean los correctos y, si no lo son, convertirlos. 
<br><br>
Nos damos cuenta de que la columna "date", "available" y "price" no tienen el formato que deberían. 
Date debe ser convertido al formato datetime, "price" al igual que el dataset anterior debe cambiarse a float, eliminando previamente el símbolo de dólar y eliminando el punto de separador de miles; y finalmente "available" debe ser una variable boolean.
</span>
</div>

In [44]:
# Cambiamos el formato de las fechas
calendar['date'] = pd.to_datetime(calendar.date, infer_datetime_format=True)
calendar['date'].head(3)

0   2018-01-03
1   2018-01-02
2   2018-01-01
Name: date, dtype: datetime64[ns]

In [45]:
# Quitamos el signo de dólar y cualquier separador de miles donde el precio no sea NA
calendar['price'] = calendar.loc[calendar.price.notnull()].price.apply(lambda x: re.sub('[$,]', '', x))

In [46]:
# Price se cambia a numérico
calendar['price'] = pd.to_numeric(calendar.price)
calendar['price'].head(3)

0    42.0
1    42.0
2    42.0
Name: price, dtype: float64

In [47]:
# La columna 'available' se convierte en boolean
calendar['available'] = calendar['available'].map({'t': True, 'f': False})
calendar['available'].head(3)

0    True
1    True
2    True
Name: available, dtype: bool

In [48]:
calendar.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4867275 entries, 0 to 4867274
Data columns (total 4 columns):
listing_id    int64
date          datetime64[ns]
available     bool
price         float64
dtypes: bool(1), datetime64[ns](1), float64(1), int64(1)
memory usage: 116.0 MB


<div style="text-align: left;">
    <span style="color:#2467C0; font-family: 'Courier New'; font-size: 1.5em;">
Hacemos un chequeo típico de la data para verificar anormalidades o cualquier cosa que nos parezca extraña.
<br><br>
Vamos a generar una pequeña tabla con los valores mínimos y máximos y los missing values para la columna "date" y "price"
</span>
</div>

In [49]:
# Get min, max and number of missing values for Date and Price
min_date = calendar.date.min()
max_date = calendar.date.max()
num_list = len(calendar.listing_id.unique())
price_missing = len(calendar.loc[pd.isnull(calendar.price)])
date_missing = len(calendar.loc[pd.isnull(calendar.date)])
min_price = calendar.price.min()
max_price = calendar.price.max()

lista = [['Precio', min_price, max_price, price_missing], ['Fecha', min_date, max_date, date_missing]]
cols = ['Column','Min','Max','Missing']

data = pd.DataFrame(lista, columns=cols)
data


Column                  Min                  Max  Missing
0  Precio                    9                15505  2121828
1   Fecha  2017-04-07 00:00:00  2018-04-07 00:00:00        0

In [50]:
len(calendar)

4867275

<div style="text-align: left;">
    <span style="color:#2467C0; font-family: 'Courier New'; font-size: 1.5em;">
El precio parece ser muy diferente entre los apartamentos, desde 9.00 dólares por noche a 15.505.00 por noche. Recordemos que no hemos separado los pisos entre los 3 tipos de habitaciones diponibles (habitación completa, habitación compartida y piso completo) lo que podría explicar el por qué estos valores máximos y mínimos.<br><br>
Vemos también que el dataset tiene datos de un año, desde la fecha que se extrajo la data hasta un año siguiente. Nos está mostrando el calendario de reservas para un año desde el 04/07/2017.
<br><br>
Es de notar la cifra en la tabla en cuanto a datos missing de precio. Son 2121828 datos missing, que forman casi el 50% de los datos calendario completo. 
<br><br>
Los datos faltantes en este dataset, representan cuando se ha reservado un piso. 
Que un dato NA sea considerado como un piso reservado tiene lógica ya que es probable que el dato no haya podido ser extraído al momento de hacer scrapping, ya que estaba reservado. <br><br>
Convertimos este nuevo dataset limpio, a un fichero .csv
</span>
</div>

In [51]:
calendar.to_csv('./calendar_clean.csv', sep=',')

<br><div style="text-align: left;">
    <span style="font-weight: bold; color:#FF7674; font-family: 'Courier New'; font-size: 2em;">Dataset "Review":</span>
</div>

<div style="text-align: left;">
    <span style="color:#2467C0; font-family: 'Courier New'; font-size: 1.5em;">
Finalmente tenemos el dataset "Review" el cuál trabajaremos para tratar de sacar información de dicho DataSet.</span>
</div>

In [52]:
review.head(3)

listing_id         id        date  reviewer_id reviewer_name  \
0    15128601  103956852  2016-09-25     48832799        Amparo   
1    15128601  105883025  2016-10-03     87158075       Daniela   
2    15128601  106595943  2016-10-07     10102108        Fabien   

                                            comments  
0  Apartamento muy acogedor, limpio y cómodo, en ...  
1  La descripción del anuncio era correcta, las f...  
2  Nice and cosy place. Really peaceful and quiet...

In [53]:
review.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 306870 entries, 0 to 306869
Data columns (total 6 columns):
listing_id       306870 non-null int64
id               306870 non-null int64
date             306870 non-null object
reviewer_id      306870 non-null int64
reviewer_name    306870 non-null object
comments         306641 non-null object
dtypes: int64(3), object(3)
memory usage: 14.0+ MB


In [54]:
# Cambiamos el formato de las fechas
review['date'] = pd.to_datetime(review.date, infer_datetime_format=True)
review['date'].head(3)

0   2016-09-25
1   2016-10-03
2   2016-10-07
Name: date, dtype: datetime64[ns]

In [55]:
review.to_csv('./review_clean.csv', sep=',')